### PR TITLE
Feature/nccl dso

### PR DESCRIPTION
**Pull request by reyoung**
cherry-pick @dzhwinter 's commit to this PR, complete cmake file, and some unit tests.

### CODE REVIEW

File: cmake/configure.cmake [hunk 1]
```diff
@@ -68,6 +68,25 @@ else()
     if(NOT CUDNN_FOUND)
         message(FATAL_ERROR "Paddle need cudnn to compile")
     endif()
+    if (NOT NCCL_INCLUDE_DIR)
```
Comment (wangkuiyi) [inline cmake/configure.cmake]: Thanks for teaching me the logic here in our offline discussion @reyoung . I think we can shorten L71-L82 into 

```cmake
if (NOT NCCL_INCLUDE_DIR)
  message(FATAL_ERROR "Paddle needs the NCCL header file")
endif()
```

File: cmake/configure.cmake [hunk 2]
```diff
@@ -68,6 +68,25 @@ else()
...
+                    "you can just grab the nccl.h to search path, "
+                    "and enable dso support."
+                    "Even though there is no offical MacOS version of nccl, "
+                    "you can still compile nccl for MacOS by "
+                    "https://github.com/NVIDIA/nccl/pull/62")
+        else()
+            message(FATAL_ERROR "Paddle need nccl header to compile")
+        endif()
+    endif()
+    if (NOT WITH_DSO)
```
Comment (wangkuiyi) [inline cmake/configure.cmake]: It seems that we can marge the two conditions on L84 and L85 into one

```cmake
if (NOT WITH_DSO AND NOT NCCL_LIBRARY)
  message(FATAL_ERROR "Paddle needs the NCCL libraries with WITH_DSO=OFF")
endif()
```

File: cmake/nccl.cmake [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        $ENV{NCCL_ROOT} $ENV{NCCL_ROOT}/include ${CUDA_TOOLKIT_INCLUDE}
+        NO_DEFAULT_PATH)
+
+get_filename_component(__libpath_hist ${CUDA_CUDART_LIBRARY} PATH)
+
+set(TARGET_ARCH "x86_64")
+if(NOT ${CMAKE_SYSTEM_PROCESSOR})
+  set(TARGET_ARCH ${CMAKE_SYSTEM_PROCESSOR})
+endif()
+
```
Comment (wangkuiyi) [inline cmake/nccl.cmake]: Just for curiosity, could we automatically download NCCL and put the header and library files into the `build/third_party` directory?  We don't have to do this in this PR, but could just file an issue recording this idea.

File: paddle/platform/dynload/dynamic_loader.cc [hunk 4]
```diff
@@ -157,6 +162,14 @@ void GetLapackDsoHandle(void** dso_handle) {
 #endif
 }
 
+void GetNcclDsoHandle(void** dso_handle) {
```
Comment (wangkuiyi) [inline paddle/platform/dynload/dynamic_loader.cc]: Nccl => NCCL

File: paddle/platform/dynload/dynamic_loader.cc [hunk 5]
```diff
@@ -157,6 +162,14 @@ void GetLapackDsoHandle(void** dso_handle) {
...
+#if defined(__APPLE__) || defined(__OSX__)
+  GetDsoHandleFromSearchPath(FLAGS_nccl_dir, "libnccl.dylib", dso_handle);
```
Comment (wangkuiyi) [inline paddle/platform/dynload/dynamic_loader.cc]: In some future PR, we should rename Dso into DSO.

File: paddle/platform/dynload/dynamic_loader.h [hunk 6]
```diff
@@ -58,6 +58,14 @@ void GetWarpCTCDsoHandle(void** dso_handle);
  */
 void GetLapackDsoHandle(void** dso_handle);
 
+/**
+ * @brief    load the DSO of NVIDIA nccl
+ *
+ * @param    **dso_handle   dso handler
+ *
+ */
+void GetNcclDsoHandle(void** dso_handle);
```
Comment (wangkuiyi) [inline paddle/platform/dynload/dynamic_loader.h]: I am not sure, but should we rename

GetNcclDsoHandle => dso::LoadNCCL
GetLapackDsoHandle => dso::LoadLAPACK

File: paddle/platform/dynload/dynamic_loader.cc (hunk 5 above)
Comment (QiJune) [reply paddle/platform/dynload/dynamic_loader.cc]: Yes, we should use DSO for abbreviation dynamic shared object.

File: cmake/nccl.cmake (hunk 3 above)
Comment (reyoung) [reply cmake/nccl.cmake]: `nccl` will be packaged to official CUDA docker image.

We can download `nccl` header in our Paddle source code. But to keep it simple, we just assume the developers are installed CUDA and related libraries well.

From @dzhwinter 's issue, https://gitlab.com/nvidia/cuda/issues/10#note_43614309

Comment (dzhwinter) [reply cmake/nccl.cmake]: No, we shouldn't.  This PR just moves NCCL from `build/third_party` to the docker built-in environment. Partly of this PR is cherry-picked from  https://github.com/PaddlePaddle/Paddle/pull/4818, which automate download NCCL and build it locally. In #4818, we have to build it early than any other module, so we need to add NCCL as a dependency, which is not correct comparing with DSO

File: paddle/platform/dynload/dynamic_loader.cc (hunk 4 above)
Comment (reyoung) [reply paddle/platform/dynload/dynamic_loader.cc]: Done.

File: cmake/configure.cmake (hunk 1 above)
Comment (reyoung) [reply cmake/configure.cmake]: Done.

File: cmake/configure.cmake (hunk 2 above)
Comment (reyoung) [reply cmake/configure.cmake]: Done.

File: cmake/nccl.cmake (hunk 3 above)
Comment (wangkuiyi) [reply cmake/nccl.cmake]: I think I understand your point -- some users might already have a special version of NCCL installed on their systems, so PaddlePaddle should try to use it.

**Pull request merged**